### ISSUE TITLE
Help! Can't build 32-bit halite.exe. Any alternatives for offline testing?

### ISSUE THREAD

**Issue by deansgit** [HaliteChallenge/Halite-II#90]
This is related to the other Help Request for a 32-bit version of halite.exe. The only reason why I even need to create this is because all of the computers within 2 miles of me are using 32bit arch.

First: make.bat and make2017.bat reference files that don't exist

Line 6's command is
COPY .\bin\Release\x64\libzstd_static.lib ..\..\..\obj

This directory does not exist on the repo. I had to open up **MinGW**, invoke **make** in the libs directory, and then manually copy this file to the environment/obj directory.

Once I figured all of that and fixed the paths/variables, I ran make and got an LNK2001 error. After some poking around, it seems likely this is a stack overflow error. I added /Gs999999 as an option to expand the heap (per a comment on an article I read) but it didn't help. What can I do at this point? I've been working on this for three days and I'm so demotivated right now.

Here is the output from the MSVS Developer Command Prompt (VS 2017)

```
[MYPROMPT]> cl.exe /std:c++14 /Gs999999 /O2 /MT /EHsc /I . /Fo.\obj\ ^
 .\obj\libzstd_static.lib ^
 .\main.cpp ^
 .\core\Constants.cpp .\core\Entity.cpp .\core\GameEvent.cpp .\core\Halite.cpp .\core\hlt.cpp .\core\SimulationEvent.cpp ^
 .\core\Replay.cpp .\core\Statistics.cpp ^
 .\networking\Networking.cpp .\networking\BotInputError.cpp ^
 .\core\mapgen\Generator.cpp .\core\mapgen\SolarSystem.cpp .\core\mapgen\AsteroidCluster.cpp ^
 /link /ltcg /implib:obj\halite.lib /out:halite.exe

**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x86'
Microsoft (R) C/C++ Optimizing Compiler Version 19.11.25547 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

main.cpp
Constants.cpp
Entity.cpp
GameEvent.cpp
Halite.cpp
hlt.cpp
SimulationEvent.cpp
Replay.cpp
Statistics.cpp
Networking.cpp
BotInputError.cpp
Generator.cpp
SolarSystem.cpp
AsteroidCluster.cpp
Generating Code...
Microsoft (R) Incremental Linker Version 14.11.25547.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:main.exe
/ltcg
/implib:obj\halite.lib
/out:halite.exe
.\obj\libzstd_static.lib
.\obj\main.obj
.\obj\Constants.obj
.\obj\Entity.obj
.\obj\GameEvent.obj
.\obj\Halite.obj
.\obj\hlt.obj
.\obj\SimulationEvent.obj
.\obj\Replay.obj
.\obj\Statistics.obj
.\obj\Networking.obj
.\obj\BotInputError.obj
.\obj\Generator.obj
.\obj\SolarSystem.obj
.\obj\AsteroidCluster.obj
libzstd_static.lib(huf_compress.o) : error LNK2001: unresolved external symbol _
__chkstk_ms
libzstd_static.lib(fse_compress.o) : error LNK2001: unresolved external symbol _
__chkstk_ms
libzstd_static.lib(fse_decompress.o) : error LNK2001: unresolved external symbol
 ___chkstk_ms
libzstd_static.lib(fse_compress.o) : error LNK2001: unresolved external symbol _
__udivdi3
libzstd_static.lib(fse_compress.o) : error LNK2001: unresolved external symbol _
__divdi3
halite.exe : fatal error LNK1120: 3 unresolved externals
```

**Comment by Snaar** [comment]
Note how bat files reference x64 directory - that's for the 64 bit build, so yeah, that directory doesn't exist if you try to build for 32 bits.

Link errors that you have are most probably because you are mixing compilation with mingw and msvc and it's very difficult to get that to work in general - you should compile zstd with msvc as well.

Also, setting /Gs is very likely a wrong thing to do, it almost always points to something else being broken.

I'll prioritize #49 to be next thing I work on.

**Comment by Snaar** [comment]
Ok, decided to just do it right now - just pushed make2017_32.bat - please try that. Builds 32 bit using 2017 community version.

**Comment by deansgit** [comment]
Thanks! Unfortunately, I still got the same link error. Not sure what went wrong so I deleted everything and started from scratch (literally downloaded the Halite-II-master repo and then ran the make2017_32.bat file.). No dice.

Is there something else I should have done beforehand? Sorry, I'm not used to doing any of these things...

**Comment by Snaar** [comment]
Can you copy-paste the log after running make2017_32.bat?

**Comment by deansgit** [comment]
Error code was "The system cannot find the path specified" but I just figured that problem out. I had to correct the path for MSbuild in build.generic.cmd (the folder on my machine doesn't have "(x86)" in the name.

Now I have a new problem but this one seems straightforward. We'll see how far I get after installing windows sdk 8.1

D:\Dean\Projects\Halite-II-master\environment\zstd-1.3.0\build\VS_scripts>build.
vs2017.32
### Building VS2017 project for Release Win32 (v141)...
### Build Params: /verbosity:minimal /nologo /t:Clean,Build /p:Platform=Win32 /p
:Configuration=Release /p:PlatformToolset=v141 /p:OutDir="\Dean\Projects\Halite-
II-master\environment\zstd-1.3.0\build\VS_scripts\bin/Release/Win32/"
C:\Program Files\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTarge
ts\Platforms\Win32\PlatformToolsets\v141\Toolset.targets(34,5): error MSB8036:
The Windows SDK version 8.1 was not found. Install the required version of Wind
ows SDK or change the SDK version in the project property pages or by right-cli
cking the solution and selecting "Retarget solution". [D:\Dean\Projects\Halite-
II-master\environment\zstd-1.3.0\build\VS2010\libzstd\libzstd.vcxproj]

**Comment by Snaar** [comment]
Ah yeah, this looks healthier.

**Comment by Snaar** [comment]
By the way, can you tell me what path did you end up specifying?

**Comment by deansgit** [comment]
Changed it from:
"%programfiles(x86)%\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\MSBuild.exe"

to:
"%programfiles%\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\MSBuild.exe"

**Comment by deansgit** [comment]
So I finally built zstd by doing the following things:

I downloaded the Windows SDK v 8.1. Ran it. Kept getting errors that certain header files weren't found. Got lazy trying to figure out exactly which ones were missing from MSVC so I copied all of the ucrt (Universal CRT) files to MSVC includes.

All of the header files (excluding the sys folder) from here:
C:\Program Files\Windows Kits\10\Include\10.0.16299.0\ucrt
to here:
C:\Program Files\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.11.25503\include

Smooth sailing from there. Finally got a 32bit halite.exe!!! Thanks for all of your help.
